File: src/ArkadeHeroes.Web/Components/Motion.cs
using Microsoft.JSInterop;

namespace ArkadeHeroes.Web.Components;

/// <summary>
/// Whether the player has asked their system to suppress motion.
///
/// <para>CSS answers this for anything that animates. This exists for the part of a cinematic that is made
/// of WAITING rather than moving — the wave-by-wave dungeon crawls, which are paced with awaits in C#. A
/// reduced-motion player who still had to sit through those would get all of the delay and none of the
/// payoff, so they skip straight to the resolved state instead.</para>
///
/// <para>Asked once and cached: the query can change under you (an OS setting flipped mid-session), but
/// re-asking costs an interop hop per beat of an animation loop to catch a case a page reload already
/// fixes. A browser that cannot answer gets the animated path — the same default the stylesheet gives it.</para>
/// </summary>
public static class Motion
{
    private static Task<IJSObjectReference>? _module;
    private static bool? _reduced;

    public static async ValueTask<bool> PrefersReducedAsync(IJSRuntime js)
    {
        if (_reduced is bool known) return known;
        try
        {
            _module ??= js.InvokeAsync<IJSObjectReference>("import", "./js/motion.js").AsTask();
            _reduced = await (await _module).InvokeAsync<bool>("prefersReducedMotion");
        }
        catch
        {
            // A blocked module load or a browser without matchMedia is not a reason to change how the page
            // behaves — fall back to "motion is fine", and stop asking.
            _reduced = false;
        }
        return _reduced.Value;
    }
}

File: src/ArkadeHeroes.Web/wwwroot/js/motion.js
// One question, asked of the browser instead of guessed at: has this player asked for motion to stop?
//
// The stylesheet already answers it for anything that MOVES — every animation in app.css is switched off
// under `prefers-reduced-motion: reduce`. What a stylesheet cannot switch off is TIME. The dungeon crawls
// are sequenced in C# with awaits, so under reduced motion a player would still be made to sit through
// several silent seconds while a rail they cannot see moving fills itself in. This lets those sequences
// skip to the resolved state instead, which is what "reduce" actually asks for.
//
// The C# side (Motion) invokes this by name. There is no compiler between the two, so a rename here breaks
// the interop at runtime rather than at build time — MotionInteropTests pins the name.

export function prefersReducedMotion() {
    return typeof window !== 'undefined'
        && typeof window.matchMedia === 'function'
        && window.matchMedia('(prefers-reduced-motion: reduce)').matches === true;
}
